elixir/apps/web/lib/web/live/sign_up.ex

```diff
@@ -0,0 +1,234 @@
+                  <%= @account.name %>
+                </td>
+              </tr>
+              <tr>
+                <td class={~w[border-b border-slate-100 dark:border-slate-700
+                              p-4 pl-8 text-gray-900 dark:text-white]}>
+                  Account Slug:
+                </td>
+                <td class={~w[border-b border-slate-100 dark:border-slate-700
+                              p-4 pl-8 text-gray-900 dark:text-white]}>
+                  <%= @account.slug %>
+                </td>
+              </tr>
+            </tbody>
+          </table>
+        </div>
+      </div>
+      <div class="text-base leading-7 text-center text-gray-900 dark:text-white">
+        <div>
+          Sign In URL
+        </div>
+        <div>
+          <.link
+            class="font-medium text-blue-600 dark:text-blue-500 hover:underline"
+            navigate={~p"/#{@account.slug}/sign_in"}
+          >
+            <%= "#{Web.Endpoint.url()}/#{@account.slug}/sign_in" %>
+          </.link>
+        </div>
+      </div>
+    </div>
+    """
+  end
+
+  def sign_up_form(assigns) do
+    ~H"""
+    <h3 class="text-center text-m font-bold leading-tight tracking-tight text-gray-900 sm:text-xl dark:text-white">
+      Sign Up Now
+    </h3>
+    <.simple_form for={@form} class="space-y-4 lg:space-y-6" phx-submit="submit" phx-change="validate">
+      <.inputs_for :let={account} field={@form[:account]}>
+        <.input
+          field={account[:name]}
+          type="text"
+          label="Account Name"
+          placeholder="Enter an Account Name here"
+          required
+          phx-debounce="300"
+        />
+        <.input field={account[:slug]} type="hidden" />
+      </.inputs_for>
+
+      <.inputs_for :let={actor} field={@form[:actor]}>
+        <.input
+          field={actor[:name]}
+          type="text"
+          label="Your Name"
+          placeholder="Enter your name here"
+          required
+          phx-debounce="300"
+        />
+        <.input field={actor[:type]} type="hidden" />
+      </.inputs_for>
+
+      <.input
+        field={@form[:email]}
+        type="text"
+        label="Email"
+        placeholder="Enter your email here"
+        required
+        phx-debounce="300"
+      />
+
+      <:actions>
+        <.button phx-disable-with="Creating Account..." class="w-full">
+          Create Account
+        </.button>
+      </:actions>
+      <p class="text-xs text-center">
+        By signing up you agree to our <.link
+          href="https://www.firezone.dev/terms"
+          class="text-blue-600 dark:text-blue-500 hover:underline"
+        >Terms of Use</.link>.
+      </p>
+    </.simple_form>
+    """
+  end
+
+  def handle_event("validate", %{"registration" => attrs}, socket) do
+    changeset =
+      %Registration{}
+      |> Web.Registration.changeset(attrs)
+      |> Map.put(:action, :validate)
+
+    socket = assign(socket, form: to_form(changeset))
```

Comment:
I think the unique slug you generated when the page was generated will be reset here and during the submit call too.

But we should not generate it on `mount` in the first place, because there is no guarantee that it will become unavailable after the page is loaded. Instead, we should use `Ecto.Changeset.prepare_changes/3` to put a default slug value with a guarantee that it will be generated within a DB transaction.

Reply:
From my testing, I believe the slug does not get reset here.

I generate that slug in the `mount` function just as a placeholder so that the `set_default_slug` (which will become `put_default_slug`) won't run the function to make an unnecessary query to the DB to generate a unique slug after every change in the form.

I intentionally did not use the slug generated from the `mount` function in the `handle_event("save")` function, because I didn't want to allow people to potentially pass in their own slug secretly through the hidden slug field in the form and I knew there was a chance that generated slug might be taken by the time the user submitted the form.

I wasn't aware of the `Ecto.Changeset.prepare_changes/3` though, so I will take a look at that.

Reply:
As discussed offline, I've wrapped the `put_default_slug` in a call to `prepare_changes`.